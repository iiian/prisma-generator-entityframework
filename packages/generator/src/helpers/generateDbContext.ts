import { DMMF } from '@prisma/generator-helper';
import { readFileSync } from 'fs';
import * as Handlebars from 'handlebars';
import { join } from 'path';

export type GenerateDbContextParams = {
  clientClassName: string;
  namespace: string;
  connectionString: string;
  dbHost: string;
  models: DMMF.Model[];
};

Handlebars.registerHelper('onConfiguringBody', (params: GenerateDbContextParams) => {
  if (params.dbHost.toLowerCase() !== 'mysql') {
    return `optionsBuilder.Use${params.dbHost}(@"${params.connectionString}");`;
  }
  return `var connection_string = @"${params.connectionString}";
        optionsBuilder.Use${params.dbHost}(connection_string, ServerVersion.AutoDetect(connection_string));`;
});

Handlebars.registerHelper('getOnModelCreatingMethod', (params: GenerateDbContextParams) => {
  const primary_key_entities = params.models
    .filter(model => model.primaryKey?.fields)
    .map(model => ({ model, fields: model.primaryKey!.fields }));

  if (!primary_key_entities.length) {
    return '';
  }

  return `
    protected override void OnModelCreating(ModelBuilder modelBuilder) {
      ${primary_key_entities.map(entity => `modelBuilder.Entity<${entity.model.name}>()
        .HasKey(e => new { ${entity.fields.map(field => `e.${field}`).join(', ')} });`)}
    }
`;
});

export function generateDbContext(params: GenerateDbContextParams) {
  const template_text = readFileSync(join(__dirname, '..', 'templates', 'DbContext.cs.hbs')).toString();
  const template = Handlebars.compile(template_text);
  return template(params);
}