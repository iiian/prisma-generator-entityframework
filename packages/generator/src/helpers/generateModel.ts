import { DMMF } from '@prisma/generator-helper';
import { readFileSync } from 'fs';
import * as Handlebars from 'handlebars';
import { join } from 'path';

export type GenerateModelParams = {
  namespace: string;
  model: DMMF.Model;
  schema_file_path: string;
};

Handlebars.registerHelper('isForeignKey', (field: DMMF.Field, fields: DMMF.Field[], options) => {
  if (fields.some(some => some.relationFromFields?.[0] === field.name)) {
    return options.fn(this);
  }
  return options.inverse(this);
});

Handlebars.registerHelper('getForeignKey', (field: DMMF.Field, fields: DMMF.Field[]) => {
  return fields.find(some => some.relationFromFields?.[0] === field.name)!.name;
});

Handlebars.registerHelper('isTruthy', function (value) {
  return value !== undefined && value !== null;
});

Handlebars.registerHelper('getDbName', (model: DMMF.Model) => {
  return model.dbName ?? model.name;
});

Handlebars.registerHelper('getCSType', (field: DMMF.Field) => {
  let type: string;
  switch (field.type) {
    case 'Int': type = 'int'; break;
    case 'BigInt': type = 'long'; break;
    case 'Boolean': type = 'bool'; break;
    case 'Decimal':
    case 'Float': type = 'float'; break;
    case 'String': type = ' string'; break;
    default: type = field.type as string; break;
  }
  if (field.isList) {
    type = `ICollection<${type}>`;
  }
  return type;
});

Handlebars.registerHelper('getColumnAnnotation', (field: DMMF.Field, model: DMMF.Model, raw_schema_lines: string[]) => {
  const field_name = getMappedFieldName(field.name, model.name, raw_schema_lines);
  const key = isPrimaryKey(field, model) && 'Key';
  const required = field.isRequired && 'Required';
  const column_parameter_list = (field_name && `("${field_name}")` || '');
  const column = `Column${column_parameter_list}`;

  return `[${[key, required, column].filter(Boolean).join(', ')}]`;
});

function isPrimaryKey(field: DMMF.Field, model: DMMF.Model) {
  if (field.isId) {
    return true;
  }
  if (!model.primaryKey?.fields) {
    return false;
  }
  const index = model.primaryKey!.fields.findIndex(each_primary_key => each_primary_key === field.name);
  return index >= 0;
}

function getMappedFieldName(field_name: string, model_name: string, raw_schema_lines: string[]) {
  // required in order to differentiate fields w/ the same name on different models, which _could_ in theory be mapped to different
  // field names on the db side.
  const model_line_start = raw_schema_lines.findIndex(line => line.includes(`model ${model_name}`));

  // get the first line on the model that matches the field
  const line = raw_schema_lines.find((line, index) => new RegExp(`^\\s*${field_name}.*$`).test(line) && index > model_line_start)!;

  // extract the @map("field") annotation, or default to the field_name
  const db_field_name = /\@map\(\"(?<db_field_name>.*)\"\)/.exec(line)?.groups?.db_field_name ?? null;

  return db_field_name;
}

export function generateModel({ model, namespace, schema_file_path }: GenerateModelParams): string {
  const template_text = readFileSync(join(__dirname, '..', 'templates', 'Model.cs.hbs')).toString();
  const template = Handlebars.compile(template_text);
  const raw_schema_lines = readFileSync(schema_file_path).toString().split(/[\r]?\n/).filter(Boolean);
  return template({ namespace, model, raw_schema_lines });
}