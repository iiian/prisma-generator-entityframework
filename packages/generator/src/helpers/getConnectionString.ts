import { DataSource } from '@prisma/generator-helper';
import { logger } from '@prisma/sdk';
import { join } from 'path';

export function getConnectionString(datasource: DataSource): string {
  if (!datasource) {
    throw new Error('No datasource specified. I need a datasource in order to construct the DbContext. The DbContext will need to setup the connection in the `OnConfiguring` life cycle hook.');
  }
  switch (datasource.provider) {
    case 'mysql':
    case 'postgres' as any:
    case 'postgresql':
      return getMysqlOrPostgresqlConnectionString(datasource);
    case 'sqlite':
      return getSqliteConnectionString(datasource);
    case 'sqlserver':
      return getSqlServerConnectionString(datasource);
    default:
      throw new Error(`Unrecognized or unsupported datasource provider ${datasource.provider}`);
  }
}

function getMysqlOrPostgresqlConnectionString(datasource: DataSource): string {
  const conn_str_fmt: RegExp = /\w+:\/\/((?<user>.*)\:(?<password>.*)@)?(?<host>.*)\:(?<port>\d{4,5})(\/(?<initial_db>.*))?/;
  const raw_conn_str = getUrl(datasource);
  const matches = conn_str_fmt.exec(raw_conn_str);
  if (!matches) {
    throw new Error(`Connection string could not be parsed: ${raw_conn_str}`);
  }
  return `Host=${matches!.groups!.host!};Database=${matches!.groups!.initial_db!};Username=${matches!.groups!.user!};Password=${matches!.groups!.password!}`;
}

function getSqliteConnectionString(datasource: DataSource): string {
  const conn_str_fmt: RegExp = /file\:(?<db_path>.*)/;
  const raw_conn_str = getUrl(datasource);
  const matches = conn_str_fmt.exec(raw_conn_str);
  if (!matches) {
    throw new Error(`Connection string could not be parsed: ${raw_conn_str}`);
  }
  return `Data Source=${matches!.groups!.db_path!};`;
}

function getSqlServerConnectionString(datasource: DataSource): string {
  const host_fmt: RegExp = /^sqlserver:\/\/(?<host>.*)(\:(?<port>\d{4,5}))?\;(database|Database|DATABASE)\=(?<initial_db>.*)\;(user|User|USER)\=(?<user>.*)\;(password|Password|PASSWORD)\=(?<password>.*)\;(encrypt|Encrypt|ENCRYPT)\=(?<encrypted>(true|false))(\;)?$/;
  const raw_conn_str = getUrl(datasource);
  const matches = host_fmt.exec(raw_conn_str);
  if (!matches) {
    throw new Error(`Connection string could not be parsed: ${raw_conn_str}`);
  }
  return `Host=${matches!.groups!.host!};Database=${matches!.groups!.initial_db!};Username=${matches!.groups!.user!};Password=${matches!.groups!.password!};Encrypt=${matches!.groups!.encrypted!}`;
}

function getUrl(datasource: DataSource): string {
  return (datasource.url.fromEnvVar)
    ? process.env[datasource.url.fromEnvVar]!
    : datasource.url.value
    ;
}