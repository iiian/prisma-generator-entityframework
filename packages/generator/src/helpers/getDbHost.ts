import { SupportedConnector } from 'src/types';

/**
 * @param connector the serial identifier used by Prisma to associate a schema with a provider
 * @returns an entityframework mapped name to be used with a `Use<Provider>(conn_str)` statement in the DbContext subclass.
 */
export function getDbHost(connector: SupportedConnector): string {
  if (!connector) {
    throw new Error('No datasource specified. I need a datasource in order to construct the DbContext. The DbContext will need to setup the connection in the `OnConfiguring` life cycle hook.');
  }

  switch (connector) {
    case 'sqlserver':
      return 'SqlServer';
    case 'mysql':
      return 'MySql';
    case 'postgresql':
      return 'Npgsql'; // the official microsoft entity framework documentation lists Npgsql as the preferred db provider for postgres
    case 'sqlite':
      return 'Sqlite';
  }
}
