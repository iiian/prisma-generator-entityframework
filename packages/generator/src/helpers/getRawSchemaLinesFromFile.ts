import { readFileSync } from 'fs';

export function getRawSchemaLinesFromFile(schema_file_path: string) {
  return readFileSync(schema_file_path).toString().split(/[\r]?\n/).filter(Boolean);
}
