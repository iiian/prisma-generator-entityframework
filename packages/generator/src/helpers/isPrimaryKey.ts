import { DMMF } from '@prisma/generator-helper';

export function isPrimaryKey(field: DMMF.Field, model: DMMF.Model) {
  if (field.isId) {
    return true;
  }
  if (!model.primaryKey?.fields) {
    return false;
  }
  const index = model.primaryKey!.fields.findIndex(each_primary_key => each_primary_key === field.name);
  return index >= 0;
}
