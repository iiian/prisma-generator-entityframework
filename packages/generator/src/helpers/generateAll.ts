
import { DMMF } from '@prisma/generator-helper';
import { generateDbContext } from './generateDbContext';
import { generateModel } from './generateModel';

export type GenerationResult = {
  name: string;
  text: string;
};

export type GenerateAllResult = {
  client: GenerationResult;
  models: Array<GenerationResult>;
};

export type GenerateAllParams = {
  namespace: string;
  clientClassName: string;
  connectionString: string;
  dbHost: string;
  datamodel: DMMF.Datamodel;
};

export function generateAll(params: GenerateAllParams): GenerateAllResult {
  const client = {
    name: `${params.clientClassName}.cs`,
    text: generateDbContext({
      ...params,
      models: params.datamodel.models
    })
  };

  const models = params.datamodel.models.map(model => ({
    name: `${model.name}.cs`,
    text: generateModel({
      namespace: params.namespace,
      model,
    })
  }));

  return { client, models };
};