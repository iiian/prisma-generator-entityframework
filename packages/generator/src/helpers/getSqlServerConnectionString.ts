import { DataSource } from '@prisma/generator-helper';
import { getUrl } from './getConnectionString';

export function getSqlServerConnectionString(datasource: DataSource): string {
  const config: { [key: string]: string; } = {};
  const raw_conn_str = getUrl(datasource);
  const [db, ...conn_props] = raw_conn_str.split(';');
  config.host = db.replace('sqlserver://', '');

  for (const prop of conn_props) {
    let [key, value] = prop.split('=');
    key = key.replace(/[ _]/, '');
    config[key.toLowerCase()] = value;
  }
  let conn_str;
  if (config.hasOwnProperty('integratedsecurity')) {
    const db = config.initialcatalog ?? config.database;
    const trust_cert = config.trustedcertificate;

    conn_str = `Server=${config.host};`;
    conn_str += (db && `Database=${db};`) || '';
    conn_str += `Integrated Security=${config.integratedsecurity};`;
    conn_str += (trust_cert && `Trust Server Certificate=${trust_cert};`) || '';
  } else {
    const db = config.initialcatalog ?? config.database;
    const user = config.user ?? config.username;
    const pass = config.pass ?? config.password;
    const encrypt = config.encrypt ?? config.encrypted;

    conn_str = `Host=${config.host};`;
    conn_str += (db && `Database=${db};`) || '';
    conn_str += (user && `Username=${user};`) || '';
    conn_str += (pass && `Password=${pass};`) || '';
    conn_str += (encrypt && `Encrypt=${encrypt};`) || '';
  }
  return conn_str;
  // const host_fmt: RegExp = /^sqlserver:\/\/(?<host>.*)(\:(?<port>\d{4,5}))?\;(database|Database|DATABASE)\=(?<initial_db>.*)\;(user|User|USER)\=(?<user>.*)\;(password|Password|PASSWORD)\=(?<password>.*)\;(encrypt|Encrypt|ENCRYPT)\=(?<encrypted>(true|false))(\;)?$/;
  // const raw_conn_str = getUrl(datasource);
  // const matches = host_fmt.exec(raw_conn_str);
  // if (!matches) {
  //   const backup_fmt = /^sqlserver:\/\/(?<host>.*)(\:(?<port>\d{4,5}))?\;(database|Database|DATABASE)\=(?<initial_db>.*)\;((trusted|Trusted|TRUSTED)[ _](connection|Connection|CONNECTION))\=(?<trusted_connection>(true|True|TRUE|false|False|FALSE))/;
  //   const backup_matches = backup_fmt.exec(raw_conn_str);
  //   if (!backup_matches) {
  //     throw new Error(`Connection string could not be parsed: ${raw_conn_str}`);
  //   }
  //   return `Server=${backup_matches!.groups!.host!};Database=${backup_matches!.groups!.initial_db!};Trusted_Connection=${backup_matches!.groups!.trusted_connection!}`;
  // }
  // return `Host=${matches!.groups!.host!}:${matches!.groups!.port!};Database=${matches!.groups!.initial_db!};Username=${matches!.groups!.user!};Password=${matches!.groups!.password!};Encrypt=${matches!.groups!.encrypted!}`;
}