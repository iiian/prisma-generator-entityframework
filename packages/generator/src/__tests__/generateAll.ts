import { generateAll } from '../helpers/generateAll';
import { getTestDMMF } from './__fixtures__/getTestDMMF';

describe('generateAll', () => {
  it('should invoke generateModel for each model', async () => {
    const { dmmf: sampleDMMF, sample_prisma_schema_path } = await getTestDMMF();

    expect(generateAll({
      clientClassName: 'ExampleClient',
      namespace: 'ProcessingFramework',
      dbHost: 'Npgsql',
      schema_file_path: sample_prisma_schema_path,
      connectionString: 'Host=my_postgres_host.com;Database=initial_db;Username=user;Password=password;',
      datamodel: sampleDMMF.datamodel,
    })).toMatchSnapshot();
  });
});
