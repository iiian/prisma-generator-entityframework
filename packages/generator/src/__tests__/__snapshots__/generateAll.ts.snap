// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateAll should invoke generateModel for each model 1`] = `
Object {
  "client": Object {
    "name": "ExampleClient.cs",
    "text": "using Microsoft.EntityFrameworkCore;

namespace ProcessingFramework {
  public class ExampleClient: DbContext {
    public DbSet<User> User { get; set; }
    public DbSet<Post> Post { get; set; }
    public DbSet<Comment> Comment { get; set; }
    
    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) {
        optionsBuilder.UseNpgsql(@\\"Host=my_postgres_host.com;Database=initial_db;Username=user;Password=password;\\");
    }
  }
}",
  },
  "models": Array [
    Object {
      "name": "User.cs",
      "text": "using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace ProcessingFramework {
  [Table(\\"user\\")]
  public class User {
    [Key, Required, Column]
    public int id { get; set; }
    [Required, Column]
    public  string email { get; set; }
    [Column]
    public  string name { get; set; }
    public ICollection<Post> posts { get; set; }
    public ICollection<Comment> comment { get; set; }
  }
}",
    },
    Object {
      "name": "Post.cs",
      "text": "using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace ProcessingFramework {
  [Table(\\"post\\")]
  public class Post {
    [Key, Required, Column]
    public int id { get; set; }
    [Required, Column]
    public  string content { get; set; }
    [Required, Column(\\"created_at\\")]
    public DateTime createdAt { get; set; }
    [ForeignKey(\\"createdBy\\")]
    [Required, Column(\\"creator_id\\")]
    public int creatorId { get; set; }
    public User createdBy { get; set; }
    public ICollection<Comment> comment { get; set; }
  }
}",
    },
    Object {
      "name": "Comment.cs",
      "text": "using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace ProcessingFramework {
  [Table(\\"comment\\")]
  public class Comment {
    [Key, Required, Column]
    public int id { get; set; }
    public Post post { get; set; }
    public User creator { get; set; }
    [ForeignKey(\\"post\\")]
    [Required, Column(\\"post_id\\")]
    public int postId { get; set; }
    [ForeignKey(\\"creator\\")]
    [Required, Column(\\"creator_id\\")]
    public int creatorId { get; set; }
  }
}",
    },
  ],
}
`;
